DR2-1559 Add representationType and representationSuffix

MancunianSam opened the pull request:
These shouldn't be null so we're validating them.

MancunianSam committed:
DR2-1559 Add representationType and representationSuffix

These shouldn't be null so we're validating them.

techncl commented on src/main/scala/uk/gov/nationalarchives/DynamoFormatters.scala:
```diff
@@ -182,4 +188,10 @@ object DynamoFormatters {
 
   case object File extends Type
 
+  sealed trait RepresentationType
```
Should we import this from the Preservica client?

MancunianSam replied:
I would like to because otherwise it's a pain seperating the dependencies with the same name like in the reconciler. The problem is that there's no dependency on the Preservica client in the dynamo formatters and it doesn't feel like there should be either. 
I think the best we can do is rename these so they don't clash with the Preservica ones. `FileRepresentationType` or something like that. What do you think?

techncl replied:
Hmm, but they are both representation types of files. is there an adjective that could differentiate the two e.g. "{X}RepresentationType"? If not, then we can just use `FileRepresentationType`.

MancunianSam replied:
They are both representation types of files so I think `FileRepresentationType` is ok.

MancunianSam replied:
I'll change the case objects I think to `RepresentationTypePreservation` and `RepresentationTypeAccess` to avoid clashes as well I think.

techncl replied:
Maybe the other way around? like `PreservationRepresentationType` and `AccessRepresentationType`?

MancunianSam replied:
Yeah, that seems sensible, I've changed them.

techncl reviewed (commented):
Looks good! Just 1 question, as well as:

1. The type for `representationSuffix` is a `String`, shouldn't it be a `Number`?
2. I just remembered that Kurtis said that the `representationSuffix` is an index (and I think Preservica refer to it as such); should we call it `representationIndex` or `representationTypeIndex`? It's only used as a suffix in OCFL

MancunianSam commented:
It is used as a suffix in the asset opex creator as well but now I think about it, I think it would be better as a number and calling it `representationTypeIndex` 
This forces the representation name to be `Preservation_1` `Access_2` so if anyone wants `Access_Primary` then they can't but I think a bit of consistency is no bad thing.
I'll change it to use a numeric field.

MancunianSam committed:
Change suffix to Int

Also, fixed the write method for the File table as it wasn't working and
added a test for it.

MancunianSam committed:
Rename fields

MancunianSam committed:
Add toString method

MancunianSam committed:
Ran scalafmt

techncl commented on src/main/scala/uk/gov/nationalarchives/DynamoWriteUtils.scala:
```diff
@@ -57,8 +57,8 @@ object DynamoWriteUtils {
           representationType -> DynamoValue.fromString(fileDynamoTable.representationType match {
-            case DynamoFormatters.RepresentationTypePreservation => "Preservation"
-            case DynamoFormatters.RepresentationTypeAccess       => "Access"
+            case DynamoFormatters.PreservationRepresentationType => "Preservation"
```
Now that you've added a `toString`, Do you think it makes sense to do `DynamoFormatters.PreservationRepresentationType.toString` instead of `"Preservation"`? or 

```
case presResType: DynamoFormatters.PreservationRepresentationType => presResType.ToString // I forgot whether it's ":" or "@"
```

Same question for `"Access"`, below and line 71 and 72 in the DynamoReadUtils file

MancunianSam replied:
We don't need this match statement in the WriteUtils at all any more. It can just be `DynamoValue.fromString(fileDynamoTable.representationType.toString)`

The static strings in the ReadUtils need to stay as is, the compiler complains if I try to use the toString method.

techncl reviewed (commented):
Just one more comment

MancunianSam committed:
Use toString method

techncl reviewed (approved):
👍
